### ISSUE TITLE
Incorrect Regular Expression in `/lib/build.js`

### ISSUE THREAD

**Issue by PatOnTheBack** [Modernizr/Modernizr#2460]
Line 42 and 43 of `/lib/build.js` says:

```JavScript
// remove last return statement and trailing })
contents = contents.replace(/return.*[^return]*$/, '');
```

[lgtm spotted a duplicate r](https://lgtm.com/projects/g/Modernizr/Modernizr/snapshot/8b121bd16ccc619bef5060e78eccb6420c4e7b87/files/lib/build.js?sort=name&dir=ASC&mode=heatmap) inside `[^return]`.

Should the lines be like this instead?

```JavaScript
// remove last return statement and trailing })
contents = contents.replace(/return.*([^r][^e][^t][^u][^r][^n])*$/, '');
```

**Comment by rejas** [comment]
This proposed new line looks overly complicated. Why is the old one a problem?

**Comment by PatOnTheBack** [comment]
@rejas The old line is a problem because 'r' is repeated inside of the negated list `[^return]*$`. With the way the regular expression is written, the final asterisk allows the negated set to be ignored completely.
The old regular expression matches:

- "return apple"
- "return return" // This should not be matched.
- "return alert"

I don't think my proposed regular expression fixed the problem, but the regex definitely needs to be fixed. One possible way to fix it is using [negative look-arounds](https://stackoverflow.com/questions/406230/regular-expression-to-match-a-line-that-doesnt-contain-a-word).

**Comment by rejas** [comment]
Thx for the info. Would you be willing to help and to create a PR for this?

**Comment by PatOnTheBack** [comment]
@Rejas I am willing to help with a PR. To know how to best write the regex, I need to understand why the matching needed to be performed in the first place. The code was not commented, so I wasn't entirely sure what the regex was for.

**Comment by knittl** [comment]
Indeed, this regular expression looks like it was intended match any `'return'` not followed by another `'return'`. But `[^return]` does not mean "not the string 'return' at this position", but rather "not any of the characters in the group at this position".

It would be really helpful to have a list of expected input and output. As mentioned earlier, it is not exactly clear what the code is supposed to do; the assumption being "remove the everything from a line after the last 'return' in this line". Here's my attempt to define input=>output mappings:

* 'return me return not}' => 'return me'
* 'return 0;}' => ''
* 'return return;' => 'return'
* 'return {};}' => ''
* 'return apple' => ''
* 'return alert;}' => ''
* 'more return' => 'more'
* 'more return me' => 'more'
* 'more return yes return no' => 'more return yes'

Can anybody confirm or reject?

**Comment by knittl** [comment]
A better (more maintainable?) solution might be to use `.replace` with a callback function, see #2613: every line with at least one `'return'` is matched, then use a callback to remove everything between the last `'return'` occurrence in this line and the end of line. WDYT?
